File: training-data/zig/generic_queue.zig
const std = @import("std");

fn Queue(comptime T: type) type {
    return struct {
        items: std.ArrayList(T),

        const Self = @This();

        fn init(allocator: std.mem.Allocator) Self {
            return Self{ .items = std.ArrayList(T).init(allocator) };
        }

        fn enqueue(self: *Self, item: T) !void {
            try self.items.append(item);
        }

        fn dequeue(self: *Self) T {
            return self.items.orderedRemove(0);
        }
    };
}

pub fn main() !void {
    const allocator = std.heap.page_allocator;
    var q = Queue(i32).init(allocator);
    try q.enqueue(1);
    try q.enqueue(2);
    try q.enqueue(3);
    std.debug.print("{d}\n", .{q.dequeue()});
    std.debug.print("{d}\n", .{q.dequeue()});
}

File: training-data/perl/generic_queue.pl
use strict;
use warnings;

package Queue;

sub new {
    my $class = shift;
    return bless { items => [] }, $class;
}

sub enqueue {
    my ($self, $item) = @_;
    push @{$self->{items}}, $item;
}

sub dequeue {
    my $self = shift;
    return shift @{$self->{items}};
}

package main;

my $q = Queue->new();
$q->enqueue(1);
$q->enqueue(2);
$q->enqueue(3);
print $q->dequeue(), "\n";
print $q->dequeue(), "\n";
